feat: handle Windows file paths with backslashes in URL parsing

mertcanaltin opened the pull request:
fyi @annevk for https://github.com/whatwg/url/issues/873


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/url/874.html" title="Last updated on Jun 28, 2025, 9:15 PM UTC (1e2f500)">Preview</a> | <a href="https://whatpr.org/url/874/59979a1...1e2f500.html" title="Last updated on Jun 28, 2025, 9:15 PM UTC (1e2f500)">Diff</a>

annevk commented on url.bs:
```diff
@@ -2247,6 +2247,18 @@ and then runs these steps:
+       <li><p>Set <var>url</var>'s <a for=url>scheme</a> to "<code>file</code>".
+       <li><p>Set <var>buffer</var> to the empty string.
+       <li><p>Replace every U+005C (\) code point in <a>remaining</a> with U+002F (/).
+       <li><p>Prepend "<code>///</code>" to <a>remaining</a>.
+       <li><p>Set <var>state</var> to <a>file state</a>.
```
Is there a difference between this and immediately jumping to the path state?

mertcanaltin replied:
thank you yes this is simpler and cleaner, this is how I update

annevk commented on url.bs:
```diff
@@ -2247,6 +2247,18 @@ and then runs these steps:
+      <ol>
+       <li><p>Set <var>url</var>'s <a for=url>scheme</a> to "<code>file</code>".
+       <li><p>Set <var>buffer</var> to the empty string.
+       <li><p>Replace every U+005C (\) code point in <a>remaining</a> with U+002F (/).
```
Why is this needed?

annevk replied:
I think we still want to treat them as validation errors (note that those are not fatal). This entire feature is a legacy feature, it doesn't deserve to be valid or even partially valid.

mertcanaltin replied:
Thanks, I Updated to emit invalid-reverse-solidus validation errors for each backslash while still doing the replacement for legacy compatibility.

annevk commented on url.bs:
```diff
@@ -2247,6 +2247,18 @@ and then runs these steps:
      append <a>c</a>, <a lt="ASCII lowercase">lowercased</a>, to <var>buffer</var>.
 
+     <li>
+      <p>Otherwise, if <a>c</a> is U+003A (:) and <var>buffer</var>'s <a for=string>length</a> is 1 and 
+      <var>buffer</var> contains only an <a>ASCII alpha</a> and <a>remaining</a> starts with U+005C (\), then:
```
```suggestion
      <p>Otherwise, if <a>c</a> is U+003A (:), <var>buffer</var>'s <a for=string>length</a> is 1, 
      <var>buffer</var> contains only an <a>ASCII alpha</a>, and <a>remaining</a> starts with
      U+005C (\):
```

annevk commented on url.bs:
```diff
@@ -2247,6 +2247,18 @@ and then runs these steps:
+      <var>buffer</var> contains only an <a>ASCII alpha</a> and <a>remaining</a> starts with U+005C (\), then:
+
+      <ol>
+       <li><p>Set <var>url</var>'s <a for=url>scheme</a> to "<code>file</code>".
```
A newline after each `<li>` except the last one.

annevk reviewed (commented):
This looks surprisingly straightforward. Have you perhaps implemented this in jsdom as well to see what tests would need adjusting? @domenic would you care to have a look?

cc @jasnell @achristensen07 @anonrig

domenic commented:
Please restore the pull request template that you deleted.

Would you mind performing the following workflow to validate your spec changes vs. the test changes?

- Clone https://github.com/jsdom/whatwg-url
- Make local edits to https://github.com/jsdom/whatwg-url/blob/main/lib/url-state-machine.js that follow the spec exactly. (Don't worry about performance or anything, just try to get it exactly following the spec changes you made.)
- Make local edits to https://github.com/jsdom/whatwg-url/blob/main/scripts/get-latest-platform-tests.js to update the commit hash to the latest hash from https://github.com/web-platform-tests/wpt/pull/53459 . (Currently `b2f42a872f99c6b2876f90d04d5a28ed86219fbd`.)
- Run `npm install` then `npm test`.

mertcanaltin commented:
> Please restore the pull request template that you deleted.
> 
> Would you mind performing the following workflow to validate your spec changes vs. the test changes?
> 
> * Clone https://github.com/jsdom/whatwg-url
> * Make local edits to https://github.com/jsdom/whatwg-url/blob/main/lib/url-state-machine.js that follow the spec exactly. (Don't worry about performance or anything, just try to get it exactly following the spec changes you made.)
> * Make local edits to https://github.com/jsdom/whatwg-url/blob/main/scripts/get-latest-platform-tests.js to update the commit hash to the latest hash from [accepting windows backslash paths web-platform-tests/wpt#53459](https://github.com/web-platform-tests/wpt/pull/53459) . (Currently `b2f42a872f99c6b2876f90d04d5a28ed86219fbd`.)
> * Run `npm install` then `npm test`.

Thanks, I tried and some test get fail, I will try solved

tests that fail:
https://gist.github.com/mertcanaltin/64ea116998228b15c47513f24143aa3e

annevk commented:
@mertcanaltin heya, you'll need to restore the PR template as requested by Domenic. It would also be good to have an update on jsdom/whatwg-url and the test failure you ran into. Editorially this PR looks fine, modulo the double newline, but we want to have evidence that it's good with running code and tests.

mertcanaltin commented:
> @mertcanaltin heya, you'll need to restore the PR template as requested by Domenic. It would also be good to have an update on jsdom/whatwg-url and the test failure you ran into. Editorially this PR looks fine, modulo the double newline, but we want to have evidence that it's good with running code and tests.

thank you, unfortunately I could not find the pr template, I will look at it again immediately, thank you very much

mertcanaltin commented:
in addition I think for this PR I need to open an issue in 
Chromium: ...
Gecko: ...
WebKit: ...
Deno: ...
Node.js: ... 
projects, right

mertcanaltin commented:
I opened issues related repos, now I will fixed test problem @annevk

mertcanaltin commented:
@domenic @annevk I've implemented this in jsdom/whatwg-url and
  updated the test expectations:

  **Implementation PR:** jsdom/whatwg-url#304

  **Test Results:**
  - ✅ **5353/5381 tests passing (99.5%)**
  - ✅ All basic Windows path scenarios working:
    - Drive letter paths: `C:\path\file.txt` →
  `file:///C:/path/file.txt`
    - UNC paths: `\\server\share\file.txt` →
  `file://server/share/file.txt`
    - Unicode characters correctly percent-encoded
    - Backslashes converted to forward slashes

  **27 Edge Cases Need Spec Clarification:**

  The failing tests fall into these categories:

  1. **Invalid drive letters** (should these fail or be accepted?)
     - `CC:\path\file.txt`
     - `C:\\\path\file.txt` (triple backslash)
     - `C:\\` (just drive + backslashes)

  2. **Device paths** (should these be supported?)
     - `\\.\Y:`
     - `\\.\y:`

  3. **Special characters in paths:**
     - `C:\folder#fragment\file.txt` (hash in path)
     - `C:\folder%20encoded\file.txt` (percent-encoding)
     - `C:\folder\file?.txt` (question mark)
     - Paths with tabs (`\t`)

  4. **UNC paths with base URL:**
     - `\\x\hello` against `http://example.org/foo/bar`

  Could you provide guidance on the expected behavior for these edge
  cases? I'm happy to update both the spec and tests once we clarify
  the intended behavior.

  **Note:** WPT tests have been updated to use percent-encoded
  Unicode (matching actual browser behavior).
